performance/single-node/lock_tests.c

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright (c) 2016 Cray Inc. All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+/******* Tunable defines for testing different lock implementations *******/
+#define rlock_acquire  pthread_rwlock_rdlock
+#define rwlock_acquire pthread_rwlock_wrlock
+#define release        pthread_rwlock_unlock
+#define lock_t	       pthread_rwlock_t
+/**************************************************************************/
+
+#include <pthread.h>
+#include <stdlib.h>
+#include <assert.h>
+/* #include <print_utils.h> */
+#define FIELD_WIDTH 16
+#include <stdio.h>
+
+#include <ct_utils.h>
+
+static pthread_barrier_t barrier;
+static pthread_t *writers, *readers;
+static lock_t lock;
+
+/*******************************************************************************
+ * Begin test functions
+ ******************************************************************************/
+void *reader(void *arg)
+{
+	unsigned nreads = *(unsigned *)arg, i;
+	pthread_barrier_wait(&barrier);
+	START_CLOCK;
+
+	for (i = 0; i < nreads; i++) {
+		rlock_acquire(&lock);
+		/* read */
+		release(&lock);
+	}
+
+	return NULL;
+}
+
+void *writer(void *arg)
+{
+	unsigned nwrites = *(unsigned *) arg, i;
+	pthread_barrier_wait(&barrier);
+	START_CLOCK;
+
+	for (i = 0; i < nwrites; i++) {
+		rwlock_acquire(&lock);
+		/* write */
+		release(&lock);
+	}
+
+	return NULL;
+}
```

Comment:
These test do measure the speed of the lock, but they do nothing to verify the integrity of the data. Since I suspect these tests will be used to verify your rwlock implementation at some point, you should verify that the data contents inside the writer and reader regions are copacetic. Add something simple.

```C 

static int counter;

reader ()  { 
....
read_acquire()
expected = counter;
actual = counter;
release(&lock);
assert(expected == actual);

writer() 
....
write_acquire()
expected = counter +1 
counter = counter + 1
actual = counter
release()
assert(expected == actual)